packages/backend/src/modules/bridges/BridgeComparator.ts

```diff
@@ -0,0 +1,129 @@
+import type { Logger } from '@l2beat/backend-tools'
+import type { Database } from '@l2beat/database'
+import type { BridgeComparePlugin, BridgeExternalItem } from './compare/types'
+
+export class BridgeComparator {
+  private running = false
+  private pluginData: Record<
+    string,
+    {
+      items: (BridgeExternalItem & { isLatest: boolean })[]
+      type: BridgeComparePlugin['type']
+    }
+  > = {}
+
+  constructor(
+    private db: Database,
+    private plugins: BridgeComparePlugin[],
+    private logger: Logger,
+    private intervalMs = 10 * 60_000,
+  ) {
+    this.logger = logger.for(this)
+  }
+
+  start() {
+    setInterval(this.runCompare, this.intervalMs)
+    this.logger.info('Compare scheduled', {
+      intervalMs: this.intervalMs,
+    })
+    this.runCompare()
+    this.logger.info('Started')
+  }
+
+  async runCompare() {
+    if (this.running) {
+      return
+    }
+    this.running = true
+    try {
+      await this.fetchExternalItems()
+      await this.compare()
+    } catch (e) {
+      this.logger.error(e)
+    }
+    this.running = false
+  }
+
+  async fetchExternalItems() {
+    this.logger.info('Fetching items from external explorers...', {
+      plugins: this.plugins.map((p) => p.name),
+    })
+
+    await Promise.all(
+      this.plugins.map(async (plugin) => {
+        try {
+          const items = (await plugin.getExternalItems()).map((i) => ({
+            ...i,
+            isLatest: true,
+          }))
+
+          if (!this.pluginData[plugin.name]) {
+            this.pluginData[plugin.name] = {
+              type: plugin.type,
+              items: items,
+            }
+          } else {
+            this.pluginData[plugin.name].items.push(...items)
+          }
+        } catch (error) {
+          this.logger.warn(
+            `Fetching items for plugin ${plugin.name} failed`,
+            error,
+          )
+        }
+      }),
+    )
+  }
+
+  async compare() {
+    this.logger.info('Comparing...', {
+      plugins: Array.from(Object.keys(this.pluginData)),
+    })
+
+    for (const [plugin, { items, type }] of Object.entries(this.pluginData)) {
+      if (items.length === 0) return
+      if (items.length === 0) continue
+
+      const records =
+        type === 'message'
+          ? await this.db.bridgeMessage.getExistingItems(items)
+          : await this.db.bridgeTransfer.getExistingItems(items)
```

Comment:
You can simplify this by not looping over plugins. Every plugin just collects items to the same list and tags them with plugin name. Then compare just gets all items.

Furthermore you can simplify this again by modifying the db query from `(from = a and to = b) or (from = c and to = d)` to `from in (a, c) and to in (b, d)`. There should be almost no false positives.

Then instead of doing `find` you can precompute a set from the existing items by concatenating `srcTxHash` and `dstTxHash` together to form the string in the set. After you create the set it's trivial to loop over and you avoid N^2 comparisons.